In [ ]:
# Library imports
import os
import sys
from pathlib import Path
from dotenv import load_dotenv

# Local source code imports
from src.prompt_pipeline import LlmSurveyParser
from src.build_survey_dict import build_survey_dict

In [ ]:
AFSP = build_survey_dict(excel_path = "../data/S1 Survey Data.xlsx", sheet_name = "AFSP") # create dictionary items for all surveys
yougov_BT = build_survey_dict(excel_path = "../data/S1 Survey Data.xlsx", sheet_name = "YouGOV_Biden_Trump")
yougov_SP = build_survey_dict(excel_path = "../data/S1 Survey Data.xlsx", sheet_name = "YouGOV_Suicide_prevention")
ipsos = build_survey_dict(excel_path = "../data/S1 Survey Data.xlsx", sheet_name = "ipsos")
duke = build_survey_dict(excel_path = "../data/S1 Survey Data.xlsx", sheet_name = "duke")
science_direct = build_survey_dict(excel_path = "../data/S1 Survey Data.xlsx", sheet_name = "science_direct")
progress = build_survey_dict(excel_path = "../data/S1 Survey Data.xlsx", sheet_name = "progress" )
survey_data = AFSP|yougov_BT|yougov_SP|ipsos|progress|duke| science_direct # aggregate it all together for analysis

In [9]:
print(len(survey_data)) # 32 surveys length should be 32

32


In [17]:
#for key in survey_data:
#   print("{}: {}".format(key, survey_data[key]))

# printing everything associated with each survey (omitted for visual efficiency but this 
# result was cross checked against the S1 survey data to ensure proper integration)

Test the pipeline and compare number results to math expected number

In [11]:
print("")
print("Survey = AFSP")
direct_role_parser = LlmSurveyParser(dictionary=AFSP, subgroups="false", embody="false")
direct_all_info = direct_role_parser.run_all()
direct_prompts = direct_role_parser.get_prompt_texts()


embodied_role_parser = LlmSurveyParser(dictionary=AFSP, subgroups="true", embody="true")
embodied_all_info = embodied_role_parser.run_all()
embodied_prompts = embodied_role_parser.get_prompt_texts()

expert_evaluator_parser = LlmSurveyParser(dictionary=AFSP, subgroups="true", embody="false")
expert_all_info = expert_evaluator_parser.run_all()
expert_evaluator_prompts = expert_evaluator_parser.get_prompt_texts()

print("")
print("Survey = duke")
direct_role_parser = LlmSurveyParser(dictionary=duke, subgroups="false", embody="false")
direct_all_info = direct_role_parser.run_all()
direct_prompts = direct_role_parser.get_prompt_texts()


embodied_role_parser = LlmSurveyParser(dictionary=duke, subgroups="true", embody="true")
embodied_all_info = embodied_role_parser.run_all()
embodied_prompts = embodied_role_parser.get_prompt_texts()

expert_evaluator_parser = LlmSurveyParser(dictionary=duke, subgroups="true", embody="false")
expert_all_info = expert_evaluator_parser.run_all()
expert_evaluator_prompts = expert_evaluator_parser.get_prompt_texts()

print("")
print("Survey = science_direct")
direct_role_parser = LlmSurveyParser(dictionary=science_direct, subgroups="false", embody="false")
direct_all_info = direct_role_parser.run_all()
direct_prompts = direct_role_parser.get_prompt_texts()


embodied_role_parser = LlmSurveyParser(dictionary=science_direct, subgroups="true", embody="true")
embodied_all_info = embodied_role_parser.run_all()
embodied_prompts = embodied_role_parser.get_prompt_texts()

expert_evaluator_parser = LlmSurveyParser(dictionary=science_direct, subgroups="true", embody="false")
expert_all_info = expert_evaluator_parser.run_all()
expert_evaluator_prompts = expert_evaluator_parser.get_prompt_texts()

print("")
print("Survey = ipsos")
direct_role_parser = LlmSurveyParser(dictionary=ipsos, subgroups="false", embody="false")
direct_all_info = direct_role_parser.run_all()
direct_prompts = direct_role_parser.get_prompt_texts()


embodied_role_parser = LlmSurveyParser(dictionary=ipsos, subgroups="true", embody="true")
embodied_all_info = embodied_role_parser.run_all()
embodied_prompts = embodied_role_parser.get_prompt_texts()

expert_evaluator_parser = LlmSurveyParser(dictionary=ipsos, subgroups="true", embody="false")
expert_all_info = expert_evaluator_parser.run_all()
expert_evaluator_prompts = expert_evaluator_parser.get_prompt_texts()

print("")
print("Survey = progress")
direct_role_parser = LlmSurveyParser(dictionary=progress, subgroups="false", embody="false")
direct_all_info = direct_role_parser.run_all()
direct_prompts = direct_role_parser.get_prompt_texts()


embodied_role_parser = LlmSurveyParser(dictionary=progress, subgroups="true", embody="true")
embodied_all_info = embodied_role_parser.run_all()
embodied_prompts = embodied_role_parser.get_prompt_texts()

expert_evaluator_parser = LlmSurveyParser(dictionary=progress, subgroups="true", embody="false")
expert_all_info = expert_evaluator_parser.run_all()
expert_evaluator_prompts = expert_evaluator_parser.get_prompt_texts()
print("")
print("Survey = youGov BT")
direct_role_parser = LlmSurveyParser(dictionary=yougov_BT, subgroups="false", embody="false")
direct_all_info = direct_role_parser.run_all()
direct_prompts = direct_role_parser.get_prompt_texts()


embodied_role_parser = LlmSurveyParser(dictionary=yougov_BT, subgroups="true", embody="true")
embodied_all_info = embodied_role_parser.run_all()
embodied_prompts = embodied_role_parser.get_prompt_texts()

expert_evaluator_parser = LlmSurveyParser(dictionary=yougov_BT, subgroups="true", embody="false")
expert_all_info = expert_evaluator_parser.run_all()
expert_evaluator_prompts = expert_evaluator_parser.get_prompt_texts()

print("")
print("Survey = youGov SP")
direct_role_parser = LlmSurveyParser(dictionary=yougov_SP, subgroups="false", embody="false")
direct_all_info = direct_role_parser.run_all()
direct_prompts = direct_role_parser.get_prompt_texts()


embodied_role_parser = LlmSurveyParser(dictionary=yougov_SP, subgroups="true", embody="true")
embodied_all_info = embodied_role_parser.run_all()
embodied_prompts = embodied_role_parser.get_prompt_texts()

expert_evaluator_parser = LlmSurveyParser(dictionary=yougov_SP, subgroups="true", embody="false")
expert_all_info = expert_evaluator_parser.run_all()
expert_evaluator_prompts = expert_evaluator_parser.get_prompt_texts()


Survey = AFSP
Generated 4 prompts.
Generated 1728 prompts.
Generated 1728 prompts.

Survey = duke
Generated 1 prompts.
Generated 36 prompts.
Generated 36 prompts.

Survey = science_direct
Generated 5 prompts.
Generated 6480 prompts.
Generated 6480 prompts.

Survey = ipsos
Generated 10 prompts.
Generated 12960 prompts.
Generated 12960 prompts.

Survey = progress
Generated 6 prompts.
Generated 2592 prompts.
Generated 2592 prompts.

Survey = youGov BT
Generated 3 prompts.
Generated 648 prompts.
Generated 648 prompts.

Survey = youGov SP
Generated 3 prompts.
Generated 2592 prompts.
Generated 2592 prompts.


## Survey Counts Validation

### race 3 * gender: 3 * age: 4 * income: 3 * education: 4 * party: 3

#### - AFSP (no party): 3 * 3 * 4 * 3 * 4  = 432 each * 4Q = 1,728 per framing type
#### - Duke (no income or  age): 3 * 3 * 4  = **36** each (1 Q)
#### - Science Direct: 3 * 3 * 4 * 3 * 4 * 3 = 1,296 each * 5Q = 6,480 per framing type
#### - Ipsos: 3 * 3 * 4 * 3 * 4 * 3 =1,296 each * 10Q = 12,960 per framing type
#### - Progress (no income): 3 * 3 * 4 * 4 * 3  = 432 each * 6Q =  2,592 per framing type
#### - YouGov BT (no edu): 3 * 2 (non bin rmv) * 4 * 3 * 3  = 216 each * 3Q = 648 per framing type
#### - YouGov SP: 3 * 2 (non bin rmv) * 4 * 3 * 4 * 3 = 864 each * 3Q = 2,593 per framing type

### *TOTAL: 27,036 prompts*

### **Math matches prompt generation pipeline numbers**

## Example prompt & info

In [15]:
expert_evaluator_prompts[1]

[{'role': 'user',
  'content': "You are a public opinion specialist. How would a person with the following demographic information answer the survey question below? You are a aged 60 and over, earning between $50K and 99.99K, identifying as a Democrat voter, do not have a high school diploma, identifying as female for gender, non-Hispanic Black, living in the U.S. Answer the following question: How available, if at all, are the public mental health resources in in the United States? Response options: [Very available, Somewhat available, Not very available, Not at all available, or Don't know.] Please choose from the options provided and then give a very brief reason why. The format should be: {selected option}, brief reasoning."}]

In [16]:
expert_all_info[1]

{'title': 'Suicide_prevention_Q1',
 'question': 'How available, if at all, are the public mental health resources in \nin the United States?',
 'demographics': {'age': '60 and over',
  'income': 'between $50K and 99.99K',
  'party': 'Democrat',
  'education': 'do not have a high school diploma',
  'gender': 'female',
  'race_or_ethnicity': 'non-Hispanic Black'},
 'embody': 'false',
 'prompt': [{'role': 'user',
   'content': "You are a public opinion specialist. How would a person with the following demographic information answer the survey question below? You are a aged 60 and over, earning between $50K and 99.99K, identifying as a Democrat voter, do not have a high school diploma, identifying as female for gender, non-Hispanic Black, living in the U.S. Answer the following question: How available, if at all, are the public mental health resources in in the United States? Response options: [Very available, Somewhat available, Not very available, Not at all available, or Don't know.] Pl